flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetSingleRowJoin.scala

```diff
@@ -138,32 +146,97 @@ class DataSetSingleRowJoin(
     val condition = codeGenerator.generateExpression(joinCondition)
 
     val joinMethodBody =
-      s"""
-        |${condition.code}
-        |if (${condition.resultTerm}) {
-        |  ${conversion.code}
-        |  ${codeGenerator.collectorTerm}.collect(${conversion.resultTerm});
-        |}
-        |""".stripMargin
+      if (joinType == JoinRelType.INNER) {
+        s"""
+         |${condition.code}
+         |if (${condition.resultTerm}) {
+         |  ${conversion.code}
+         |  ${codeGenerator.collectorTerm}.collect(${conversion.resultTerm});
+         |}
+         |""".stripMargin
+    } else {
+        val singleNode =
+          if (rightIsSingle) {
+            rightNode
+          }
+          else {
+            leftNode
+          }
+
+        val notSuitedToCondition = singleNode
+          .getRowType
+          .getFieldList
+          .map(field => getRowType.getFieldNames.indexOf(field.getName))
+          .map(i => s"${conversion.resultTerm}.setField($i,null);")
+
+        if (joinType == JoinRelType.LEFT && leftIsSingle) {
+          s"""
+             |${condition.code}
+             |${conversion.code}
+             |if(!${condition.resultTerm}){
+             |${notSuitedToCondition.mkString("\n")}
+             |}
+             |if(!${condition.leftNullTerm}){
+             |${codeGenerator.collectorTerm}.collect(${conversion.resultTerm});
+             |}
+             |""".stripMargin
+        } else if (joinType == JoinRelType.RIGHT && rightIsSingle){
+          s"""
+             |${condition.code}
+             |${conversion.code}
+             |if(!${condition.resultTerm}){
+             |${notSuitedToCondition.mkString("\n")}
+             |}
+             |if(!${condition.leftNullTerm} && ${condition.resultTerm}){
+             |${codeGenerator.collectorTerm}.collect(${conversion.resultTerm});
+             |}
+             |""".stripMargin
+        } else {
+          s"""
+             |${condition.code}
+             |${conversion.code}
+             |if(!${condition.resultTerm}){
```

Comment:
we can add `|| ${codeGenerator.input1Term} == null` to check if the first input is null for the `JoinRelType.RIGHT && leftIsSingle` case and `|| ${codeGenerator.input2Term} == null` for the `JoinRelType.LEFT && rightIsSingle` case